services/extended_sqs_client/extended_sqs_client.go

```diff
@@ -0,0 +1,338 @@
+}
+
+func NewExtendedSQSClient(sqs aws_sqsiface.SQSAPI, config *ExtendedSQSClientConfiguration) *ExtendedSQSClient {
+	payloadStore := payload_store.NewPayloadStore(config.s3, config.s3BucketName)
+
+	return &ExtendedSQSClient{
+		sqsClient:    sqs,
+		config:       config,
+		payloadStore: payloadStore,
+	}
+}
+
+func (c *ExtendedSQSClient) SendMessage(input *aws_sqs.SendMessageInput) (*aws_sqs.SendMessageOutput, error) {
+	if input == nil {
+		errorMessage := "sendMessageRequest cannot be null."
+
+		return nil, errors.SDKError{Message: errorMessage}
+	}
+
+	if !c.config.IsPayloadSupportEnabled() {
+		return c.sqsClient.SendMessage(input)
+	}
+
+	if input.MessageBody == nil {
+		errorMessage := "messageBody cannot be null or empty."
+
+		return nil, errors.SDKError{Message: errorMessage}
+	}
+
+	if err := c.checkMessageAttributes(input.MessageAttributes); err != nil {
+		return nil, err
+	}
+
+	var sqsInput *aws_sqs.SendMessageInput
+
+	if c.config.IsAlwaysThroughS3() || c.isLarge(input) {
+		var err error
+		sqsInput, err = c.storeMessageInS3(input)
+
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		sqsInput = input
+	}
+
+	return c.sqsClient.SendMessage(sqsInput)
+}
+
+func (c *ExtendedSQSClient) ReceiveMessage(input *aws_sqs.ReceiveMessageInput) (*aws_sqs.ReceiveMessageOutput, error) {
+	if input == nil {
+		errorMessage := "receiveMessageRequest cannot be null."
+
+		return nil, errors.SDKError{Message: errorMessage}
+	}
+
+	if !c.config.IsPayloadSupportEnabled() {
+		return c.sqsClient.ReceiveMessage(input)
+	}
+
+	reservdAttributeName := sqs_configs_constants.RESERVED_ATTRIBUTE_NAME
+	legacyReservedAttributeName := sqs_configs_constants.LEGACY_RESERVED_ATTRIBUTE_NAME
+	var updatedMessageAttributeNames []*string
+	for _, name := range input.MessageAttributeNames {
+		if *name != reservdAttributeName && *name != legacyReservedAttributeName {
+			copied_name := *name
+			updatedMessageAttributeNames = append(updatedMessageAttributeNames, &copied_name)
+		}
+	}
+	updatedMessageAttributeNames = append(updatedMessageAttributeNames, &reservdAttributeName)
+	updatedMessageAttributeNames = append(updatedMessageAttributeNames, &legacyReservedAttributeName)
+
+	updatedInput := &aws_sqs.ReceiveMessageInput{}
+	*updatedInput = *input
+	updatedInput.MessageAttributeNames = updatedMessageAttributeNames
+
+	output, err := c.sqsClient.ReceiveMessage(updatedInput)
+	if err != nil {
+		return nil, err
+	}
+
+	messages := output.Messages
+	modifiedMessages := make([]*aws_sqs.Message, len(messages))
+
+	for index, message := range messages {
+		modifiedMessage := &aws_sqs.Message{}
+		*modifiedMessage = *message
+
+		messageAttributes := message.MessageAttributes
+		largePayloadAttributeName := getReservedAttributeNameIfPresent(messageAttributes)
+		if largePayloadAttributeName != nil {
+			messagePointer := strings.Replace(*message.Body, "com.amazon.sqs.javamessaging.MessageS3Pointer", "software.amazon.payloadoffloading.PayloadS3Pointer", 1)
```

Comment:
can we skip this? seems not too useful

Reply:
depends on whether you want to support java legacy sdk version. If not, can skip this

Reply:
removed